File: frontend/web/src/realtime/reducer.ts
// Realtime reducer (T-305): snapshot + events -> company state, as pure functions.
//
// This is the TypeScript twin of backend/autora/realtime/reducer.py; both must turn the same
// snapshot and events into the same projection (checked against the fixture the Python
// contract test writes: src/realtime/__fixtures__/contract.json). The rules, briefly:
//
// - Events with seq <= lastSeq, without seq, or of another company are ignored (duplicates at
//   the backlog/live boundary). The gateway guarantees order and completeness per connection
//   (05 §4 as amended in T-303), so no gap detection happens here.
// - AGENT_CREATED adds an agent (with the department it was hired into), AGENT_ASSIGNED moves
//   it to another role and room, AGENT_RETIRED removes one. Activity events set stored_state from the event type, detail
//   = payload + run/task/workflow context (none for IDLE and PAUSED; task_name is the task's
//   display name), since changes only when the state does. A non-final AGENT_RUN_FAILED or
//   AGENT_RUN_ABORTED is trace data and changes nothing.
// - TASK_CREATED adds a task (PENDING); TASK_* set its state, since and lastEventSeq;
//   TASK_STARTED also sets attempt, run and agent.
// - view(now): effective activity state (COMPLETED past display_until reads IDLE), finished tasks
//   older than 10 minutes left out, the last 100 events. Those tasks also leave the state itself
//   as later events arrive (T-412: the store must not grow with history).
//
// No visual state lives here (05 §5): poses, bubbles and animation are derived elsewhere.
import { parseEvent, type EventEnvelope } from "@autora/event-schema";

import type {
  ActivityState,
  ActivityView,
  RealtimeSnapshot,
  TaskView,
} from "./snapshot";

export const FINISHED_TASK_WINDOW_MS = 10 * 60 * 1000;
export const RECENT_EVENTS_VIEW = 100;
export const RECENT_EVENTS_KEPT = 500;

export interface LiveProgress {
  runId: string | null;
  stepSeq: number;
  tokensSoFar: number;
  progress: { label: string; current: number; target: number | null } | null;
  at: string;
}

export interface AgentState {
  id: string;
  role: string;
  display_name: string;
  avatar_key: string;
  department_id: string | null;
  department_key: string | null;
  /** null until the agent's first activity event (right after AGENT_CREATED). */
  activity: ActivityView | null;
  /** Latest ephemeral AGENT_STEP_PROGRESS of the agent's current run; never persisted. */
  liveProgress: LiveProgress | null;
}

export interface RealtimeState {
  companyId: string;
  lastSeq: number;
  agents: Record<string, AgentState>;
  tasks: Record<string, TaskView>;
  recentEvents: EventEnvelope[];
}

/** What load_snapshot returns: the comparable projection at a moment. */
export interface Projection {
  company_id: string;
  last_seq: number;
  agents: (Omit<AgentState, "liveProgress" | "activity"> & {
    activity: ActivityView;
  })[];
  tasks: TaskView[];
  recent_events: EventEnvelope[];
}

const ACTIVITY_BY_TYPE: Partial<
  Record<EventEnvelope["event_type"], ActivityState>
> = {
  AGENT_IDLE: "IDLE",
  AGENT_RESUMED: "IDLE",
  AGENT_THINKING: "THINKING",
  AGENT_WORKING: "WORKING",
  AGENT_WAITING: "WAITING",
  AGENT_REVIEWING: "REVIEWING",
  AGENT_RUN_COMPLETED: "COMPLETED",
  AGENT_RUN_FAILED: "FAILED",
  AGENT_RUN_ABORTED: "FAILED",
  AGENT_PAUSED: "PAUSED",
};
const RUNLESS: ReadonlySet<ActivityState> = new Set(["IDLE", "PAUSED"]);
const TASK_STATE_BY_TYPE: Partial<Record<EventEnvelope["event_type"], string>> =
  {
    TASK_READY: "READY",
    TASK_STARTED: "RUNNING",
    TASK_WAITING: "WAITING_APPROVAL",
    TASK_SUCCEEDED: "SUCCEEDED",
    TASK_CANCELLED: "CANCELLED",
    TASK_BLOCKED: "BLOCKED_BUDGET",
  };
const FINISHED: ReadonlySet<string> = new Set([
  "SUCCEEDED",
  "FAILED",
  "CANCELLED",
]);

// --- hydrate --------------------------------------------------------------------------------

export function hydrate(snapshot: RealtimeSnapshot): RealtimeState {
  const agents: Record<string, AgentState> = {};
  for (const agent of snapshot.agents) {
    agents[agent.id] = {
      ...agent,
      activity: { ...agent.activity },
      liveProgress: null,
    };
  }
  const tasks: Record<string, TaskView> = {};
  for (const task of snapshot.tasks) tasks[task.id] = { ...task };
  const recentEvents: EventEnvelope[] = [];
  for (const raw of snapshot.recent_events) {
    const parsed = parseEvent(raw);
    if (parsed.ok) recentEvents.push(parsed.event);
  }
  return {
    companyId: snapshot.company_id,
    lastSeq: snapshot.last_seq,
    agents,
    tasks,
    recentEvents,
  };
}

// --- apply ----------------------------------------------------------------------------------

/**
 * Apply one persisted event. Returns the same object when the event is ignored.
 *
 * Tasks that finished more than FINISHED_TASK_WINDOW_MS before the event are dropped from the
 * state (04 §7: the store must not grow with history; view() already leaves them out, so the
 * projection is unchanged). Measured from the event's time, not the local clock, so replaying
 * the same events gives the same state.
 */
export function applyEvent(
  state: RealtimeState,
  event: EventEnvelope,
): RealtimeState {
  const next = applyOne(state, event);
  return next === state
    ? state
    : pruneFinishedTasks(next, Date.parse(event.occurred_at));
}

function pruneFinishedTasks(state: RealtimeState, at: number): RealtimeState {
  let tasks: Record<string, TaskView> | null = null;
  for (const [id, task] of Object.entries(state.tasks)) {
    if (
      FINISHED.has(task.state) &&
      Date.parse(task.since) <= at - FINISHED_TASK_WINDOW_MS
    ) {
      tasks ??= { ...state.tasks };
      delete tasks[id];
    }
  }
  return tasks ? { ...state, tasks } : state;
}

function applyOne(state: RealtimeState, event: EventEnvelope): RealtimeState {
  if (
    event.seq === null ||
    event.seq <= state.lastSeq ||
    event.company_id !== state.companyId
  ) {
    return state;
  }
  const next: RealtimeState = {
    ...state,
    lastSeq: event.seq,
    recentEvents: [...state.recentEvents, event].slice(-RECENT_EVENTS_KEPT),
  };

  if (event.event_type === "AGENT_CREATED" && event.agent_id) {
    next.agents = {
      ...state.agents,
      [event.agent_id]: {
        id: event.agent_id,
        role: event.payload.role,
        display_name: event.payload.display_name,
        avatar_key: event.payload.avatar_key,
        department_id: event.payload.department_id ?? null,
        department_key: event.payload.department_key ?? null,
        activity: null,
        liveProgress: null,
      },
    };
    return next;
  }

  if (
    event.event_type === "AGENT_ASSIGNED" &&
    event.agent_id &&
    state.agents[event.agent_id]
  ) {
    // the only event that moves a drawn agent to another room without a full reload
    next.agents = {
      ...state.agents,
      [event.agent_id]: {
        ...state.agents[event.agent_id],
        role: event.payload.role,
        department_id: event.payload.department_id,
        department_key: event.payload.department_key ?? null,
      },
    };
    return next;
  }

  if (event.event_type === "AGENT_RETIRED" && event.agent_id) {
    const { [event.agent_id]: gone, ...rest } = state.agents; // off the roster
    void gone;
    next.agents = rest;
    return next;
  }

  const activityState = ACTIVITY_BY_TYPE[event.event_type];
  if (activityState && event.agent_id && state.agents[event.agent_id]) {
    const payload = event.payload as { final?: boolean };
    const traceOnly =
      (event.event_type === "AGENT_RUN_FAILED" ||
        event.event_type === "AGENT_RUN_ABORTED") &&
      payload.final === false;
    if (!traceOnly) {
      const agent = state.agents[event.agent_id];
      next.agents = {
        ...state.agents,
        [agent.id]: withActivity(agent, activityState, event, state),
      };
    }
    return next;
  }

  if (event.event_type === "TASK_CREATED" && event.task_id) {
    next.tasks = {
      ...state.tasks,
      [event.task_id]: {
        id: event.task_id,
        name: event.payload.name,
        display_name: event.payload.display_name,
        required_role: event.payload.required_role,
        state: "PENDING",
        depends_on: [...event.payload.depends_on],
        workflow_run_id: event.payload.workflow_run_id,
        attempt: 0,
        run_id: null,
        agent_id: null,
        since: event.occurred_at,
        last_event_seq: event.seq,
      },
    };
    return next;
  }

  if (
    event.event_type.startsWith("TASK_") &&
    event.task_id &&
    state.tasks[event.task_id]
  ) {
    next.tasks = {
      ...state.tasks,
      [event.task_id]: withTaskEvent(state.tasks[event.task_id], event),
    };
  }
  return next;
}

function withActivity(
  agent: AgentState,
  state: ActivityState,
  event: EventEnvelope,
  current: RealtimeState,
): AgentState {
  const runless = RUNLESS.has(state);
  const taskName =
    runless || !event.task_id
      ? null
      : (current.tasks[event.task_id]?.display_name ?? null);
  const context: Record<string, string | null> = {
    run_id: runless ? null : event.run_id,
    task_id: runless ? null : event.task_id,
    workflow_run_id: runless ? null : event.workflow_run_id,
    task_name: taskName,
  };
  const detail: Record<string, unknown> = {
    ...(event.payload as Record<string, unknown>),
  };
  for (const [key, value] of Object.entries(context))
    if (value) detail[key] = value;

  const previous = agent.activity;
  const activity: ActivityView = {
    state,
    stored_state: state,
    detail,
    since:
      previous && previous.stored_state === state
        ? previous.since
        : event.occurred_at,
    run_id: context.run_id,
    task_id: context.task_id,
    last_event_seq: event.seq as number,
  };
  const liveProgress =
    agent.liveProgress && agent.liveProgress.runId === activity.run_id
      ? agent.liveProgress
      : null;
  return { ...agent, activity, liveProgress };
}

function withTaskEvent(task: TaskView, event: EventEnvelope): TaskView {
  const next: TaskView = {
    ...task,
    since: event.occurred_at,
    last_event_seq: event.seq as number,
  };
  if (event.event_type === "TASK_FAILED") {
    if (event.payload.final) next.state = "FAILED";
    return next; // a retried failure is followed by TASK_READY in the same transaction
  }
  const target = TASK_STATE_BY_TYPE[event.event_type];
  if (target) next.state = target;
  if (event.event_type === "TASK_STARTED") {
    next.attempt = event.payload.attempt;
    next.run_id = event.payload.run_id;
    next.agent_id = event.payload.agent_id;
  }
  return next;
}

// --- ephemeral ------------------------------------------------------------------------------

export interface EphemeralMessage {
  event_type: string;
  agent_id: string | null;
  run_id: string | null;
  occurred_at?: string | null;
  payload: Record<string, unknown>;
}

/** Live progress (no seq, never part of the projection). Unknown kinds are ignored. */
export function applyEphemeral(
  state: RealtimeState,
  message: EphemeralMessage,
): RealtimeState {
  if (message.event_type !== "AGENT_STEP_PROGRESS" || !message.agent_id)
    return state;
  const agent = state.agents[message.agent_id];
  if (!agent) return state;
  const payload = message.payload as {
    step_seq?: number;
    tokens_so_far?: number;
    progress?: LiveProgress["progress"];
  };
  const liveProgress: LiveProgress = {
    runId: message.run_id,
    stepSeq: payload.step_seq ?? 0,
    tokensSoFar: payload.tokens_so_far ?? 0,
    progress: payload.progress ?? null,
    at: message.occurred_at ?? new Date().toISOString(),
  };
  return {
    ...state,
    agents: { ...state.agents, [agent.id]: { ...agent, liveProgress } },
  };
}

// --- view -----------------------------------------------------------------------------------

export function effectiveState(
  activity: ActivityView,
  now: Date,
): ActivityState {
  if (activity.stored_state === "COMPLETED") {
    const until = activity.detail.display_until;
    if (typeof until === "string" && Date.parse(until) <= now.getTime())
      return "IDLE";
  }
  return activity.stored_state;
}

export function isTaskVisible(task: TaskView, now: Date): boolean {
  return (
    !FINISHED.has(task.state) ||
    Date.parse(task.since) > now.getTime() - FINISHED_TASK_WINDOW_MS
  );
}

/** The projection at `now`: what the snapshot endpoint would return at that moment. */
export function view(state: RealtimeState, now: Date): Projection {
  const agents: Projection["agents"] = [];
  for (const agent of Object.values(state.agents)) {
    if (!agent.activity) continue;
    agents.push({
      id: agent.id,
      role: agent.role,
      display_name: agent.display_name,
      avatar_key: agent.avatar_key,
      department_id: agent.department_id,
      department_key: agent.department_key,
      activity: {
        ...agent.activity,
        state: effectiveState(agent.activity, now),
      },
    });
  }
  return {
    company_id: state.companyId,
    last_seq: state.lastSeq,
    agents,
    tasks: Object.values(state.tasks).filter((task) =>
      isTaskVisible(task, now),
    ),
    recent_events: state.recentEvents.slice(-RECENT_EVENTS_VIEW),
  };
}

File: backend/autora/realtime/reducer.py
"""Reference reducer: snapshot + events -> snapshot (T-302, 3d-office/05 §3).

The browser store (T-305) applies events to a hydrated snapshot. This module is the executable
specification of that reduction, in Python, next to the projection it must agree with:

    view(from_snapshot(load_snapshot(t0)) + events(t0, t1], now) == load_snapshot(t1, now)

``tests/realtime/test_contract.py`` checks it on randomised runtime histories (real events from
the real task manager, approvals and activity service), and exports a fixture that the
TypeScript reducer must reproduce too.

Rules (each mirrors the code that writes the row the projection reads):

- **Duplicates**: an event with ``seq <= last_seq`` is ignored. Ephemeral events (no seq) never
  change the projection.
- **Agents**: AGENT_CREATED adds the agent. An activity event sets ``stored_state`` from its type
  (``autora.runtime.activity``), ``detail`` = payload + run/task/workflow context (cleared for
  IDLE and PAUSED; ``task_name`` is the task's display name), ``since`` changes only when the
  state does, ``last_event_seq`` = the event's seq. A non-final AGENT_RUN_FAILED or
  AGENT_RUN_ABORTED is trace data and changes nothing (the activity service rejects it too). A
  PAUSED agent stays PAUSED through task-manager transitions: no activity event is written then
  (``set_activity_unless_paused``), so there is nothing to apply.
- **Tasks**: TASK_CREATED adds the task (PENDING). TASK_READY / STARTED / WAITING / SUCCEEDED /
  FAILED(final) / CANCELLED / BLOCKED set the state; TASK_STARTED also sets attempt, run and
  agent; every TASK_* event sets ``since`` and ``last_event_seq``.
- **View** (at ``now``): effective activity state (COMPLETED past ``display_until`` reads IDLE),
  finished tasks older than the window left out, the last 100 events.

Not reproduced (not in the events): ``links`` in activity detail (no domain provides them yet),
agents created without AGENT_CREATED (only test fixtures do that).
"""

from __future__ import annotations

import uuid
from collections import deque
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

from autora.company import events as company_ev
from autora.db.models import ActivityState, TaskState
from autora.realtime.projection import (
    FINISHED_TASK_WINDOW,
    RECENT_EVENTS,
    ActivityView,
    AgentView,
    CycleView,
    RealtimeSnapshot,
    TaskView,
)
from autora.runtime.events import catalog as ev
from autora.runtime.events.schema import EventEnvelope

_ACTIVITY: dict[type, ActivityState] = {
    ev.AgentIdle: ActivityState.IDLE,
    ev.AgentResumed: ActivityState.IDLE,
    ev.AgentThinking: ActivityState.THINKING,
    ev.AgentWorking: ActivityState.WORKING,
    ev.AgentWaiting: ActivityState.WAITING,
    ev.AgentReviewing: ActivityState.REVIEWING,
    ev.AgentRunCompleted: ActivityState.COMPLETED,
    ev.AgentRunFailed: ActivityState.FAILED,
    ev.AgentRunAborted: ActivityState.FAILED,
    ev.AgentPaused: ActivityState.PAUSED,
}
_RUNLESS = (ActivityState.IDLE, ActivityState.PAUSED)
_TASK_STATE: dict[type, TaskState] = {
    ev.TaskReady: TaskState.READY,
    ev.TaskStarted: TaskState.RUNNING,
    ev.TaskWaiting: TaskState.WAITING_APPROVAL,
    ev.TaskSucceeded: TaskState.SUCCEEDED,
    ev.TaskCancelled: TaskState.CANCELLED,
    ev.TaskBlocked: TaskState.BLOCKED_BUDGET,
}
_FINISHED = {TaskState.SUCCEEDED, TaskState.FAILED, TaskState.CANCELLED}


@dataclass
class _Agent:
    id: uuid.UUID
    role: str
    display_name: str
    avatar_key: str
    department_id: uuid.UUID | None = None
    department_key: str | None = None
    activity: ActivityView | None = None


@dataclass
class RealtimeState:
    company_id: uuid.UUID
    last_seq: int = 0
    cycle: CycleView | None = None
    agents: dict[uuid.UUID, _Agent] = field(default_factory=dict)
    tasks: dict[uuid.UUID, TaskView] = field(default_factory=dict)
    recent: deque[EventEnvelope] = field(default_factory=lambda: deque(maxlen=RECENT_EVENTS))

    @classmethod
    def from_snapshot(cls, snapshot: RealtimeSnapshot) -> RealtimeState:
        state = cls(
            company_id=snapshot.company_id,
            last_seq=snapshot.last_seq,
            cycle=snapshot.cycle.model_copy() if snapshot.cycle else None,
        )
        for agent in snapshot.agents:
            state.agents[agent.id] = _Agent(
                agent.id, agent.role, agent.display_name, agent.avatar_key,
                agent.department_id, agent.department_key, agent.activity.model_copy(),
            )  # fmt: skip
        state.tasks = {task.id: task.model_copy() for task in snapshot.tasks}
        state.recent.extend(snapshot.recent_events)
        return state

    # --- events ----------------------------------------------------------------------------

    def apply(self, event: EventEnvelope) -> bool:
        """Apply one event. Returns False when it was ignored (duplicate, ephemeral, other
        company)."""
        if event.seq is None or event.seq <= self.last_seq or event.company_id != self.company_id:
            return False
        self.last_seq = event.seq
        self.recent.append(event)
        payload = event.payload

        if isinstance(payload, ev.AgentCreated) and event.agent_id is not None:
            self.agents[event.agent_id] = _Agent(
                event.agent_id,
                payload.role,
                payload.display_name,
                payload.avatar_key,
                payload.department_id,
                payload.department_key,
            )
        elif isinstance(payload, company_ev.AgentAssigned) and event.agent_id in self.agents:
            # the only thing that moves a drawn agent to another room without a reload
            agent = self.agents[event.agent_id]
            agent.role = payload.role
            agent.department_id = payload.department_id
            agent.department_key = payload.department_key
        elif isinstance(payload, company_ev.CycleStarted) and event.cycle_id is not None:
            self.cycle = CycleView(
                id=event.cycle_id,
                seq=payload.seq,
                stage=payload.stage,
                deadline=payload.deadline,
            )
        elif isinstance(payload, company_ev.CycleStageChanged) and self.cycle is not None:
            self.cycle = self.cycle.model_copy(
                update={"stage": payload.to_stage, "deadline": payload.deadline}
            )
        elif isinstance(payload, ev.AgentRetired) and event.agent_id is not None:
            self.agents.pop(event.agent_id, None)  # off the roster: the office stops drawing it
        elif type(payload) in _ACTIVITY and event.agent_id in self.agents:
            if not (
                isinstance(payload, ev.AgentRunFailed | ev.AgentRunAborted) and not payload.final
            ):
                self._activity(self.agents[event.agent_id], event)
        elif isinstance(payload, ev.TaskCreated) and event.task_id is not None:
            self.tasks[event.task_id] = TaskView(
                id=event.task_id,
                name=payload.name,
                display_name=payload.display_name,
                required_role=payload.required_role,
                state=TaskState.PENDING.value,
                depends_on=list(payload.depends_on),
                workflow_run_id=payload.workflow_run_id,
                attempt=0,
                run_id=None,
                agent_id=None,
                since=event.occurred_at,
                last_event_seq=event.seq,
            )
        elif event.event_type.startswith("TASK_") and event.task_id in self.tasks:
            self._task(self.tasks[event.task_id], event)
        return True

    def _activity(self, agent: _Agent, event: EventEnvelope) -> None:
        state = _ACTIVITY[type(event.payload)]
        runless = state in _RUNLESS
        detail: dict[str, Any] = event.payload.model_dump(mode="json")
        task = self.tasks.get(event.task_id) if event.task_id else None
        context = {
            "run_id": None if runless else event.run_id,
            "task_id": None if runless else event.task_id,
            "workflow_run_id": None if runless else event.workflow_run_id,
            "task_name": None if runless or task is None else task.display_name,
        }
        detail.update(
            {k: str(v) if isinstance(v, uuid.UUID) else v for k, v in context.items() if v}
        )
        current = agent.activity
        agent.activity = ActivityView(
            state=state,
            stored_state=state,
            detail=detail,
            since=current.since
            if current is not None and current.stored_state == state
            else event.occurred_at,
            run_id=context["run_id"],
            task_id=context["task_id"],
            last_event_seq=event.seq,
        )

    @staticmethod
    def _task(task: TaskView, event: EventEnvelope) -> None:
        payload = event.payload
        task.since = event.occurred_at
        task.last_event_seq = event.seq
        if isinstance(payload, ev.TaskFailed):
            if payload.final:
                task.state = TaskState.FAILED.value
            return  # a retried failure is followed by TASK_READY in the same transaction
        target = _TASK_STATE.get(type(payload))
        if target is not None:
            task.state = target.value
        if isinstance(payload, ev.TaskStarted):
            task.attempt = payload.attempt
            task.run_id = payload.run_id
            task.agent_id = payload.agent_id

    # --- view ------------------------------------------------------------------------------

    def view(self, now: datetime) -> RealtimeSnapshot:
        """The projection at ``now``: what ``load_snapshot(..., now=now)`` returns."""
        agents = []
        for agent in self.agents.values():
            if agent.activity is None:
                continue
            activity = agent.activity.model_copy(update={"state": _effective(agent.activity, now)})
            agents.append(
                AgentView(
                    id=agent.id,
                    role=agent.role,
                    display_name=agent.display_name,
                    avatar_key=agent.avatar_key,
                    department_id=agent.department_id,
                    department_key=agent.department_key,
                    activity=activity,
                )  # fmt: skip
            )
        tasks = [
            task.model_copy()
            for task in self.tasks.values()
            if TaskState(task.state) not in _FINISHED or task.since > now - FINISHED_TASK_WINDOW
        ]
        return RealtimeSnapshot(
            company_id=self.company_id,
            last_seq=self.last_seq,
            server_time=now,
            agents=agents,
            tasks=tasks,
            recent_events=list(self.recent),
            cycle=self.cycle.model_copy() if self.cycle else None,
        )


def _effective(activity: ActivityView, now: datetime) -> ActivityState:
    if activity.stored_state is ActivityState.COMPLETED:
        until = activity.detail.get("display_until")
        if until is not None and datetime.fromisoformat(until) <= now:
            return ActivityState.IDLE
    return activity.stored_state


def canonical(snapshot: RealtimeSnapshot) -> dict[str, Any]:
    """The comparable content of a projection: keyed by id, JSON values, no server_time."""
    data = snapshot.model_dump(mode="json")
    return {
        "company_id": data["company_id"],
        "last_seq": data["last_seq"],
        "agents": {a["id"]: a for a in data["agents"]},
        "tasks": {t["id"]: t for t in data["tasks"]},
        "recent_event_seqs": [e["seq"] for e in data["recent_events"]],
        "kpis": data["kpis"],
        "cycle": data["cycle"],
    }
